### ISSUE TITLE
hvd.init() hangs with error message help-opal-shmem-mmap.txt 

### ISSUE THREAD

**Issue by maxhgerlach** [horovod/horovod#1381]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.13.1, custom built without XLA
3. Horovod version: 0.16.4
4. MPI version: Open MPI 4.0.1
5. CUDA version: 10.0
6. NCCL version: 2.3.7-1+cuda10.0
7. Python version: 2.7.12
8. OS and version: Ubuntu 16.04.5 LTS
9. GCC version: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609

Mellanox OFED 4.4-2.0.7.0 is installed.

**Bug report:**
Recently, our custom Horovod-based training code sometimes produces an error message after each worker rank called `hvd.init()`. The processes then hang without exiting. Here is the error message:
```
--------------------------------------------------------------------------
A system call failed during shared memory initialization that should
not have.  It is likely that your MPI job will now either abort or
experience performance degradation.

  Local host:  heinzel101
  System call: open(2)
  Error:       No such file or directory (errno 2)
--------------------------------------------------------------------------
[training_host:330371] 3 more processes have sent help message help-opal-shmem-mmap.txt / sys call fail
[training_host:330371] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

We start the training like this (5 local workers in this case): `mpirun --prefix /opt/openmpi -np 5 -H training_host:5 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x NCCL_DEBUG_FILE=/dev/stderr -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib --tag-output --timestamp-output python [training_script]`.

Here are gdb backtraces for all threads for one of the workers: https://gist.github.com/maxhgerlach/7e7ccbd16714c7a070156f88fce7d291 It looks like the hang is inside `BackgroundThreadLoop()` at line 929:
```
    // No ranks were given and no communicator provided to horovod_init() so use
    // MPI_COMM_WORLD
    MPI_Comm_dup(MPI_COMM_WORLD, &(ctx.mpi_comm));
```

As in #1102 for each worker we still fork out a process via multiprocessing to do IO and preprocessing _before_ initializing Horovod or MPI in the parent process. Does the error sound as if it might be related to this?

**Comment by maxhgerlach** [comment]
We still observe this issue quite often. Might this be fixed in a more recent Horovod release?

**Comment by maxhgerlach** [comment]
In the mean time I have updated to Open MPI 4.0.2 and Horovod 0.19, but this issue still occurs with many of our training jobs. Unfortunately it is very hard to reproduce consistently.

I noticed that there are a couple of open issues regarding the Vader shared memory BTL in Open MPI, e.g. https://github.com/open-mpi/ompi/issues/7308

@alsrgv @tgaddair Would you recommend switching to `sm` instead of `vader`? Since expensive data transfers run over NCCL anyway, a slight decrease in MPI performance should not matter for us.

**Comment by maxhgerlach** [comment]
After further investigation it looks like I misattributed this issue to Horovod or Open MPI. 

A default setting in systemd, that we did not know about, causes shared memory files in `/dev/shm` to be deleted while the Horovod training making use of them is still running. See https://superuser.com/questions/1117764/why-are-the-contents-of-dev-shm-is-being-removed-automatically

This was very hard to pinpoint because this only happens when the training is run by our automated custom system, but not while a login shell of the same Linux user is open on the same host.
